---
title: "Mycobiome_venn"
author: "Catherine Purse"
date: "2024-08-15"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries
```{r packages, results='hide'}
x <- c("phyloseq","RColorBrewer", "tidyverse", "patchwork", "microViz", "microbiome", "MicEco", "DatawRappr", "extrafont")
lapply(x, require, character.only = TRUE)
```

# Colour palette
```{r}
cb_palette <- c("#ee8866", "#77aadd","#d3d3d3","#FDF28E","#8ED29B","#bbccee", "#AFC30A","#cc99cc","#C1E49C","#dd7788", "#009E73")
```

# Load data

* Load ASV/Taxa/Meta tables
```{r}
ASV <- read.csv("OTU-table.csv")
tax <- read.csv("taxonomy-table.csv")
meta <- read.csv("Metadata-file.csv")
```

* Reformat ASV table (turn ASV column into rows) and strip white space
```{r}
rownames(ASV) <- paste0("ASV", 1:nrow(ASV))
ASV <- subset(ASV, select = -c(X.NAME))
View(ASV)

#strip white space
ASV <- ASV %>%
  mutate(across(where(is.character), ~str_trim(., side = "both")))
```

* Same for taxonomy table
```{r}
#do the same for taxonomy table (need rownames to match)
rownames(tax) <- paste0("ASV", 1:nrow(tax))
tax <- subset(tax, select = -c(X.TAXONOMY))
View(tax)

#strip white space
tax[] <- lapply(tax, function(x) if(is.character(x)) gsub("\\s+", " ", x))

taxa <- unique(tax$Species)
taxa
```

* Same for metadata
```{r}
meta_df <- meta %>%
  column_to_rownames("X.NAME")
View(meta_df)

#strip white space
meta_df <- meta_df %>%
  mutate(across(where(is.character), ~str_trim(., side = "both")))
View(meta_df)
```

## Create phyloseq object
```{r}
#turn df in to matrices (required for phyloseq)
asv_mat <- as.matrix(ASV)
tax_mat <- as.matrix(tax)

#transform into phyloseq objects
ASV_ps <- otu_table(asv_mat, taxa_are_rows = TRUE)
tax_ps <- tax_table(tax_mat)
samples <- sample_data(meta_df)

#combine!
myco <- phyloseq(ASV_ps, tax_ps, samples)
myco
```

* Change labels
```{r}
# Get sample data from the phyloseq object
myco_df <- as.data.frame(sample_data(myco))

# Rename Mid as Adult
myco_df$agegroup<-gsub("Mid","Adult",as.character(myco_df$agegroup))

# Update the sample data in the phyloseq object
sample_data(myco) <- sample_data(myco_df)
```

* Prune samples <10000 reads (as recommended in QIB best practice)
```{r}
myco_10k <- prune_samples(sample_sums(myco) > 10000, myco)
myco_10k
```

* Clean data of outliers and unclassified ASVs (Kingdom Level)
```{r}
#remove NHP 17 (M497BDC) as ID presented with diarrhoea
myco_10k_noout <- subset_samples(myco_10k, sample_names(myco_10k) != "NHP17Caecum" & sample_names(myco_10k) != "NHP17PC" & sample_names(myco_10k) != "NHP17DC")

# Remove unidentified ASVs at Kingdom level
myco_10k_noout <- subset_taxa(myco_10k_noout, Kingdom != " unclassified_Root ")

# Check unique Phylum
unique_phylum <- unique(tax_table(myco_10k_noout)[, 'Phylum'])

# Print or view the unique data
print(unique_phylum)
```

* Remove taxa unclassified at Phylum level
```{r}
myco_10k_noout <- subset_taxa(myco_10k_noout, Phylum != " unclassified_Fungi " & Phylum != " unidentified ")
```

* Convert to relative abundance
```{r}
myco_10k_noout <- transform_sample_counts(myco_10k_noout, function(x) x/sum(x))

# Prune samples to include only those with >0.01% abundance
myco_p <- ps_prune(myco_10k_noout, min.abundance = 0.0001)
```


```{r}
sp_venn <- ps_venn(
  myco_p,
  group = 'agegroup',
  fraction = 0,
  weight = FALSE,
  relative = TRUE,
  plot = TRUE
)
sp_venn
```

* Export graph
```{r}
tiff(file.path(folder_path, "Species_Venn.tiff"), units="in", width=5, height=5, res=300)
plot(sp_venn)
dev.off()
```